spec-0013/index.md

```diff
@@ -0,0 +1,53 @@
+---
+title: "SPEC 13 — Recommended targets and naming conventions"
+number: 13
+date: 2024-06-05
+author:
+  - "Pamphile Roy <roy.pamphile@gmail.com>"
+  - "Matthias Bussonnier <bussonniermatthias@gmail.com>"
+  - "Jarrod Millman <millman@berkeley.edu>"
+discussion: https://discuss.scientific-python.org/t/spec-13-recommended-targets-and-naming-conventions
+endorsed-by:
+---
+
+## Description
+
+For consistency and decreased cognitive load across the ecosystem, this SPEC
+recommends naming conventions around various project aspects--such as project
+structure, repository layout, folder names, task runner and `pyproject.toml`
+targets name.
+
+From a cursory survey in the Scientific Python ecosystem, we discover some
+frustration from contributors and maintainer when moving from one project to
+another and believe that consistency will make it both easier for existing
+maintainer to contribute to project as well a decrease the confusion of new
+developers when contributing or creating new projects.
+
+There seem to be a strong consensus with preference for `docs` in favor of
+`docs`, and a preference for `tests` in favor of `test`.
+
+We will note though that the _extra_ optional dependencies on PyPI seem to favor
+`test` (present on 7573 packages) vs `tests` (2362 times).
+
+## Implementation
+
+For the tie being we will not pronounce ourselves on the optional extra
+`extra` dependency for `pyproject.toml`.
+
+For other targets and folders we recommend that:
+
+- Targets related to testing be named `tests` (and not `test`). For example
+  `spin tests`, `python dev.py tests`, `nox -s tests`.
+- Folders containing tests be names `tests`.
+- Targets related to documentations be named `docs` (and not `doc`). For example
+  `spin docs`, `make docs`, `tox -s docs`.
+- That the documentation `extra` optional dependency be named `docs` (and not
```

Comment:
I would like to think that adding a recommendation for `all` and `dev` are non-controversial, too?

Reply:
I strongly recommend no recommendation on `[dev]`. Having an "all in one" environment makes the environment hard to solve, sometimes impossible to solve. Solvers are very sensitive to the number of dependencies in an environment. There was a time when you couldn't install `black` and `tensorflow` in the same environment due to dependency conflicts. I've seen solvers choke for minutes[^1] trying to solve updates on massive "all-in-one" environments. And you never need to install your code formatter, or pre-commit, etc. in the same environment as your package, they are stand-alone _tools_!

And `[all]` sometimes means all true optional dependencies, not testing and development dependencies. Like `validate-pyproject[all]`.

`uv` is currently working on a solution for project-level "pipx"-like installs. And there is an effort to make dependency groups a standard in [PEP 735](https://peps.python.org/pep-0735/). How about avoiding any extra recommendation until PEP 735 is accepted or rejected, and we look into seeing if there's interest in changing the official `[doc]` extra recommendation to `[docs]`?

[^1]: On average. The record I've seen was 25 hours.

Reply:
OK, no strong preference for `[dev]`, but definitely having a true, user facing dependencies into `[all]`, so it should have no testing or documentation ones in there.